doc/fgpu.design.rst

```diff
@@ -597,147 +597,96 @@ channels) when
-
-Work groups
-~~~~~~~~~~~
-Each work group is responsible for producing a contiguous set of output
-samples (given by the constant :c:macro:`GROUP_OUT_SIZE`). To do so, it needs
-to load data from :c:macro:`LOAD_SIZE` input samples, which includes the extra
-samples needed to cater for the footprint of the low-pass filter.
-
-To maximise the arithmetic intensity and minimise the number of load/store
-operations, it's necessary for the kernel to hold a lot of data in registers.
-To avoid needing all the data at the same time, it has an outer loop that
-alternates between firstly, loading, decoding and mixing some data, and
-secondly, applying the low-pass filter. These two stages use different
-mappings of work items to work, and communicate through local memory.
-
-.. _ddc-load:
-
-Loading and unpacking
-~~~~~~~~~~~~~~~~~~~~~
-Initially (prior to the outer loop mentioned above), each work item loads the
-packed 10-bit samples for some number of input samples into registers (between
-them they load all :c:macro:`LOAD_SIZE` samples). To save space, these are
-unpacked only as needed. At present, this kernel only supports 10-bit samples,
-and not the other sample sizes supported by the wide-band PFB kernel.
-
-To simplify alignment, the input samples are divided
-into :dfn:`segments` of 16 consecutive samples, which consumes 20 bytes or
-five 32-bit words. The segments are distributed amongst the work items in
-round-robin fashion, so that work item :math:`i` holds segments :math:`i + jW`
-where :math:`W` is the work group size (:c:macro:`WGS` in the code). There
-won't be an equal number of segments for each work item, so some work items
-will be holding useless data.
-
-When a sample is required, it is unpacked, given the segment and position
-within the segment. The kernel is designed so that the position in the segment
-is always a compile-time constant (after loop unrolling), which means the
-necessary registers and shift amounts are also known at compile-time.
-
-To cheaply achieve sign extension, the value is first shifted to the top 10
-bits of a 32-bit signed integer, then shifted right. In standard C/C++ this
-is undefined behaviour, but CUDA implements the common behaviour of performing
-sign extension.
-
-In some cases the desired sample is split across a word boundary. CUDA
-provides a (hardware-accelerated) :dfn:`funnel-shift` intrinsic, which allows two
-words to be combined into a 64-bit word and shifted, retaining just the high
-32 bits of the result; this is ideal for our use case.
+subsampling are never actually computed.  Unlike the memory-bound PFB kernel,
+this kernel is at the boundary between being memory-bound and compute-bound
+(depending on the number of taps). The design thus needs a more efficient
+approach to decoding the packed samples.
+
+Each work-item is responsible for completely computing :math:`C` consecutive
+output values (:math:`C` is a tuning parameter), which it does
```

Comment:
Ah, I *see* what you did there (Re: my earlier clarification on mako/CUDA naming convention).
* `C` == unroll == consecutive output values.